### PR TITLE
[COST-4572] Add Django management command to cleanup AWS bills and a job for running it in the clowdapp

### PR DESCRIPTION
**Pull request by cgoodfred**
## Jira Ticket

[COST-4572](https://issues.redhat.com/browse/COST-4572)

## Description

This change will add a Django management command to remove all of the AWSCostEntryBill's since October 2023 that have a NULL payer_account_id. After completing the delete, it triggers a summary task for the provider and month that was relevant to the deleted bill. The delete leverages the `cascade_deletes` function that walks the Django model and performs the delete via SQL rather than the django ORM. This function was written by HAP before he left, he wrote an article with extra information about it: https://dev.to/redhap/efficient-django-delete-cascade-43i5 

This management command has two options. A dry-run that is the default, it will show all of the bills that would be deleted on a real run and the total count of bills that would be deleted and how many providers were checked for a bill. Running the management command with `--delete` causes the cascade delete to actually be performed and remove all of the old data. This will also output a pre-filled summary URL for each provider and month that had a bill removed and associated records cleaned up.

The PR also adds another Job to the clowdapp to run management commands. It is similar in setup to the existing job with a new variable:
- `MGMT_COMMAND`: This is the command to be run by the CJI, in this case `aws_null_bill_cleanup` 

## Testing

NOTE: the management command can be tested locally using `python koku/manage.py aws_null_bill_cleanup` in a similar flow as below without the namespace. The flow with ephemeral ensures the new job will trigger the management command in non-local environments.

1. Reserve an ephemeral namespace
2. Adjust your config slightly to set `INITIAL_INGEST_NUM_MONTHS: 5` and `RETAIN_NUM_MONTHS: 5` so there will be data going beyond October.
3. Deploy the latest version of Koku to that namespace
4. Load some AWS data in the namespace either via smoke tests or port forwarding and a source.
5. Verify the bills in the DB have a payer_account_id by finding the `koku-db` pod and connecting to the koku database. 
```
SELECT id, payer_account_id, billing_period_start FROM <schema>.reporting_awscostentrybill
```
Sample output:
```
 id | payer_account_id |  billing_period_start  
----+------------------+------------------------
  3 | 12345     | 2024-01-01 00:00:00+00
  4 | 12345     | 2023-12-01 00:00:00+00
  5 | 12345     | 2023-11-01 00:00:00+00
  6 | 12345     | 2023-10-01 00:00:00+00
  7 | 12345     | 2023-09-01 00:00:00+00
(5 rows)
```
6. Set `MGMT_COMMAND` to `aws_null_bill_cleanup` in your config and apply the changes to your namespace.
7. After the CJI pod is created and runs, we are not expecting any bills to be cleaned since they all have a payer account. Check the logs and you should see something like:
```
[2024-01-25 20:24:51,585] INFO None 1 In dry run mode (--delete not passed)
[2024-01-25 20:24:51,612] INFO None 1 DRY RUN: 0 bills would be deleted across 1 providers.
```
8. After verifying that, update the bills above and set the `payer_account_id` to `NULL` and check the bills again.
```
koku=# UPDATE reporting_awscostentrybill SET payer_account_id = NULL;
UPDATE 5

SELECT id, payer_account_id, billing_period_start FROM <schema>.reporting_awscostentrybill

 id | payer_account_id |  billing_period_start  
----+------------------+------------------------
  6 |                  | 2024-01-01 00:00:00+00
  7 |                  | 2023-12-01 00:00:00+00
  8 |                  | 2023-11-01 00:00:00+00
  9 |                  | 2023-10-01 00:00:00+00
 10 |                  | 2023-09-01 00:00:00+00
(5 rows)
```
9. Increment the `MGMT_INVOCATION` number and apply the new config to the namespace
10. When the job runs, this time you should see messages about some bills that would be deleted if it wasn't doing a dry run:
```
[2024-01-25 20:31:18,098] INFO None 1 In dry run mode (--delete not passed)
[2024-01-25 20:31:18,121] INFO None 1 bills [3] would be deleted for provider: d27052f4-b399-4845-9b9e-9e1ac2d31a7a
[2024-01-25 20:31:18,122] INFO None 1 bills [4] would be deleted for provider: d27052f4-b399-4845-9b9e-9e1ac2d31a7a
[2024-01-25 20:31:18,124] INFO None 1 bills [5] would be deleted for provider: d27052f4-b399-4845-9b9e-9e1ac2d31a7a
[2024-01-25 20:31:18,125] INFO None 1 bills [6] would be deleted for provider: d27052f4-b399-4845-9b9e-9e1ac2d31a7a
[2024-01-25 20:31:18,126] INFO None 1 DRY RUN: 4 bills would be deleted across 1 providers.
```
11. Verify the bills listed match the DB bills that are from October 2023 forward.
12. Update the `MGMT_COMMAND` to be `aws_null_bill_cleanup --delete` so the delete operation will be performed and increment the `MGMT_INVOCATION` again. Apply the new config to the namespace.
13. When the new pod is created and the job runs, you should now see messages that inform you the deletes were completed and a summary was run and the task id is shown:
```
[2024-01-25 20:32:50,005] INFO None 1 DELETING BILLS (--delete passed)
[2024-01-25 20:32:50,499] INFO None 1 Deletes completed and summary triggered for provider d27052f4-b399-4845-9b9e-9e1ac2d31a7a with start 2024-01-01 and end 2024-01-31, task_id: ec654221-d304-44f1-b5cc-c09388028b52
[2024-01-25 20:32:50,717] INFO None 1 Deletes completed and summary triggered for provider d27052f4-b399-4845-9b9e-9e1ac2d31a7a with start 2023-12-01 and end 2023-12-31, task_id: 997f1b08-8db8-46db-9e63-635908965191
[2024-01-25 20:32:50,917] INFO None 1 Deletes completed and summary triggered for provider d27052f4-b399-4845-9b9e-9e1ac2d31a7a with start 2023-11-01 and end 2023-11-30, task_id: 0341f760-3d18-4238-ae28-62adddbd6a88
[2024-01-25 20:32:51,116] INFO None 1 Deletes completed and summary triggered for provider d27052f4-b399-4845-9b9e-9e1ac2d31a7a with start 2023-10-01 and end 2023-10-31, task_id: bf899a46-6b13-4f22-94f3-3a2955ae12c0
[2024-01-25 20:32:51,117] INFO None 1 4 bills deleted across 1 providers.
```
14. After that occurs, verify the DB records and ensure that any remaining bills are outside the timeframe or do not have a null payer account.
```
SELECT id, payer_account_id, billing_period_start FROM <schema>.reporting_awscostentrybill

id | payer_account_id |  billing_period_start  
----+------------------+------------------------
 10 |                  | 2023-09-01 00:00:00+00
(1 row)
```
15. Check the priority worker logs and ensure that a summary was initiated.
## Notes

...

### CODE REVIEW

File: koku/masu/api/aws_bill_cleanup.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+from masu.processor.tasks import PRIORITY_QUEUE_XL
+from masu.processor.tasks import QUEUE_LIST
+
+
+LOG = logging.getLogger(__name__)
+REPORT_DATA_KEY = "cleanup_aws_bills Task IDs"
+
+
+@never_cache
+@api_view(http_method_names=["GET"])
```
Comment (maskarb) [inline koku/masu/api/aws_bill_cleanup.py:32]: This ought to be a GET or DELETE. The GET should list all the bills that should be deleted, then the DELETE will delete those bills.

It's unclear from this endpoint what we are really identifying as needing to be deleted.

**Comment by chambridge** [comment]
Random thought, but any thought to just doing this cleanup as a migration?

**Comment by maskarb** [comment]
> Random thought, but any thought to just doing this cleanup as a migration?

Or even use a CJI?

File: koku/masu/api/aws_bill_cleanup.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    tracing_id = str(uuid4())
+    schema_dict = defaultdict(list)
+    LOG.info(log_json(tracing_id, msg="kicking off bill cleanup tasks per provider", provider_count=len(providers)))
+    for provider in providers:
+        schema = provider.customer.schema_name
+        provider_uuid = provider.uuid
+        async_result = cleanup_aws_bills_task.s(schema, provider_uuid, bill_date, tracing_id).apply_async(
+            queue=queue_name
+        )
+        schema_dict[schema].append({"provider_uuid": str(provider_uuid), "task_id": str(async_result)})
```
Comment (samdoran) [inline koku/masu/api/aws_bill_cleanup.py:67]:
```suggestion
        schema_dict[schema].append({"provider_uuid": str(provider_uuid), "task_id": async_result.id})
```

File: koku/masu/api/aws_bill_cleanup.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    params = request.query_params
+    schema_name = params.get("schema")
+    bill_date = params.get("bill_date")
+    fallback_queue = PRIORITY_QUEUE
+    if is_customer_large(schema_name):
+        fallback_queue = PRIORITY_QUEUE_XL
+    queue_name = params.get("queue") or fallback_queue
+
+    if queue_name not in QUEUE_LIST:
+        errmsg = f"'queue' must be one of {QUEUE_LIST}."
```
[the remark this answers is not in the document]
Comment (samdoran) [reply koku/masu/api/aws_bill_cleanup.py:46]: If the list of possible values in `QUEUE_LIST` is potentially large, consider setting a threshold beyond which all the items will not be inserted into the message. That keeps it readable.

File: koku/masu/api/aws_bill_cleanup.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        return Response({"Error": errmsg}, status=status.HTTP_400_BAD_REQUEST)
+    if bill_date is None:
+        errmsg = "bill_date is a required parameter."
+        return Response({"Error": errmsg}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (samdoran) [inline koku/masu/api/aws_bill_cleanup.py:50]: Gather all the errors and return them all. This provides a better user experience if more than one parameter is missing.
```suggestion
    errors = []
    if queue_name not in QUEUE_LIST:
        errors.append(f"'queue' must be one of {QUEUE_LIST}.")

    if bill_date is None:
        errors.append("bill_date is a required parameter.")

    if errors:
        return Response([{"Error": message} for message in errors], status=status.HTTP_400_BAD_REQUEST)
```

**Comment by cgoodfred** [comment]
> > Random thought, but any thought to just doing this cleanup as a migration?
> 
> Or even use a CJI?

I'm not opposed to doing this as a CJI/migration. I'd probably lean more towards a CJI for that route over a migration. It removes the possibility of the clarity you were asking for in the GET method vs DELETE method comment if we go that route, fine with that being the tradeoff here?

File: koku/masu/api/aws_bill_cleanup.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+from api.provider.models import Provider
+from masu.processor import is_customer_large
+from masu.processor.tasks import cleanup_aws_bills as cleanup_aws_bills_task
+from masu.processor.tasks import PRIORITY_QUEUE
+from masu.processor.tasks import PRIORITY_QUEUE_XL
+from masu.processor.tasks import QUEUE_LIST
+
+
+LOG = logging.getLogger(__name__)
+REPORT_DATA_KEY = "cleanup_aws_bills Task IDs"
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/api/aws_bill_cleanup.py]: No longer applicable after shift to CJI.

File: koku/masu/api/aws_bill_cleanup.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+@renderer_classes(tuple(api_settings.DEFAULT_RENDERER_CLASSES))
+def aws_bill_cleanup(request):
+    """Cleanup AWS bills with a NULL payer account_id."""
+    params = request.query_params
+    schema_name = params.get("schema")
+    bill_date = params.get("bill_date")
+    fallback_queue = PRIORITY_QUEUE
+    if is_customer_large(schema_name):
+        fallback_queue = PRIORITY_QUEUE_XL
+    queue_name = params.get("queue") or fallback_queue
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/api/aws_bill_cleanup.py]: No longer applicable after shift to CJI.

File: koku/masu/api/aws_bill_cleanup.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/api/aws_bill_cleanup.py]: No longer applicable after shift to CJI.

File: koku/masu/api/aws_bill_cleanup.py (hunk 2 above)
Comment (cgoodfred) [reply koku/masu/api/aws_bill_cleanup.py]: No longer applicable after shift to CJI.

File: koku/masu/management/commands/aws_null_bill_cleanup.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+class Command(BaseCommand):
+    help = "Delete AWS Bills and with a NULL payer_account_id and all FK references for a given month."
+
+    def add_arguments(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "--delete",
+            action="store_true",
+            default=False,
+            help="Actually delete the cost entry bills.",
+        )
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply koku/masu/management/commands/aws_null_bill_cleanup.py:41]: I think for this job we can leave it without a specific schema, we need to clean up all of the schema anyway and if we absolutely need to cleanup a single schema individually, we can do that manually.

**Review by maskarb — APPROVED** [review]

File: koku/masu/management/commands/aws_null_bill_cleanup.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (djnakabaale) [reply koku/masu/management/commands/aws_null_bill_cleanup.py:41]: okay, that sounds good 👍🏾

**Review by maskarb — APPROVED** [review]

**Review by maskarb — APPROVED** [review]

**Pull request merged**